### PR TITLE
HPCC-35827 Add LogicalFileListFiltered filtering and field projection

### PR DESCRIPTION
**Pull request by jakesmith**
Introduces new file listing API with user-friendly filter syntax and field selection. LogicalFileList is marked DEPRECATED, users should use new function for efficiency and capability.

Features:
- Filter syntax: wildcards (owner:jsmith), numeric comparisons (size>1000000), date ranges (modified>2024-01-01), property checks (has:description), file type filtering (is:superfile, is:normal)
- Field projection to select specific result fields
- Configurable file limits with breach detection
- Overflow/underflow protection for numeric filters

Implementation:
- FileListResultFieldSource for IFieldSource pattern
- parseUserFilterSyntax converts user syntax to internal DFUQFilter
- Field validation via dfuFieldAttrPathMap
- Default fields: name, modified, size, rowcount, cluster, superfile, owner

Testing:
- 12 regression test cases covering filtering, projection, limits, edge cases

## Type of change:
- [ ] This change is a bug fix (non-breaking change which fixes an issue).
- [x] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [ ] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [ ] The commit message title makes sense in a changelog, by itself.
  - [ ] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [x] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [x] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Cloud-compatibility
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [x] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Smoketest:
- [ ] Send notifications about my Pull Request position in Smoketest queue.
- [ ] Test my draft Pull Request.

### CODE REVIEW

**Comment by ghalliday** [comment]
@jakesmith some errors due to initialiser order

**Comment by ghalliday** [comment]
Also add #ownwarning(2364, ignore);

to fix the regression suite mismatches.

File: ecllibrary/std/File.ecl [hunk 1]
```diff
@@ -225,7 +225,92 @@ EXPORT dataset(FsFilenameRecord) RemoteDirectory(varstring machineIP, varstring
...
+ * @param fields        Comma-separated list of field names to include in the result. Defaults to blank (all default fields).
+ *                      When blank or empty, returns default fields: name, modified, size, rowcount, cluster, superfile, owner.
+ *                      Available fields: name, modified, size, rowcount, cluster, superfile, owner, group, recordsize, numparts.
+ *                      Note: 'name' field is always included regardless of specification.
+ *                      Examples:
+ *                        ''                   - Default fields (backward compatible)
+ *                        'name,size'          - Only name and size
+ *                        'name,superfile,owner' - Name, superfile flag, and owner
+ * @param unknownszero  Whether to set file sizes that are unknown to zero(0) instead of minus-one (-1). Defaults to FALSE.
+ * @param foreigndali   The IP address of the foreign dali used to resolve the file.  If blank then the file is resolved
```
Comment (ghalliday) [inline ecllibrary/std/File.ecl]: Doesn't this need to be a foreign esp?

File: plugins/fileservices/fileservices.cpp [hunk 2]
```diff
@@ -2385,6 +2391,652 @@ FILESERVICES_API void FILESERVICES_CALL fsLogicalFileList(ICodeContext *ctx, siz
...
+ *
+ * Multiple filters separated by commas
+ */
+static void parseUserFilterSyntax(const char *userFilter, StringBuffer &internalFilter)
+{
+    if (isEmptyString(userFilter))
+        return;
+
+    StringArray terms;
+    terms.appendList(userFilter, ",");
```
Comment (ghalliday) [inline plugins/fileservices/fileservices.cpp:2665]: Documentation says space separated, but this looks like comma separated.

File: plugins/fileservices/fileservices.cpp [hunk 3]
```diff
@@ -2385,6 +2391,652 @@ FILESERVICES_API void FILESERVICES_CALL fsLogicalFileList(ICodeContext *ctx, siz
...
+        // Check for negation prefix
+        bool negate = (term[0] == '!');
+        if (negate)
+        {
+            term++;
+            if (isEmptyString(term))
+                throw makeStringException(-1, "Invalid filter syntax: '!' must be followed by a filter term");
+        }
+
+        bool matched = false;
```
Comment (ghalliday) [inline plugins/fileservices/fileservices.cpp]: minor: code would be simpler if this was set to true, and else sets to false, or throw the mismatch in the else.

File: plugins/fileservices/fileservices.cpp [hunk 4]
```diff
@@ -2385,6 +2391,652 @@ FILESERVICES_API void FILESERVICES_CALL fsLogicalFileList(ICodeContext *ctx, siz
...
+            else if (stricmp(fileType, "normal") == 0)
+            {
+                internalFilter.appendf("%u%c2%c3%c",
+                    DFUQFTspecial, DFUQFilterSeparator,
+                    DFUQFilterSeparator, DFUQFilterSeparator);
+                matched = true;
+            }
+            else if (stricmp(fileType, "any") == 0)
+            {
+                internalFilter.appendf("%u%c2%c1%c",
```
Comment (ghalliday) [inline plugins/fileservices/fileservices.cpp]: Easier to spot the difference if these three branches were commoned up (by setting a char to '1','2' or '3'.

File: plugins/fileservices/fileservices.cpp [hunk 5]
```diff
@@ -2385,6 +2391,652 @@ FILESERVICES_API void FILESERVICES_CALL fsLogicalFileList(ICodeContext *ctx, siz
...
+                // Parse numeric range
+                if (op[0] == '>')
+                {
+                    // field > value or field >= value
+                    __int64 minVal = _atoi64(value);
+                    if (!hasEquals)
+                    {
+                        if (minVal == I64C(0x7FFFFFFFFFFFFFFF))
+                            throw makeStringExceptionV(-1, "Invalid filter syntax: '%s' - value too large for > comparison (would overflow)", originalTerm);
+                        minVal++;
```
Comment (ghalliday) [inline plugins/fileservices/fileservices.cpp:2802]: This does not work for float types.

File: plugins/fileservices/fileservices.cpp [hunk 6]
```diff
@@ -2385,6 +2391,652 @@ FILESERVICES_API void FILESERVICES_CALL fsLogicalFileList(ICodeContext *ctx, siz
...
+                        attrName.str(), DFUQFilterSeparator,
+                        DFUQFilterSeparator, maxVal, DFUQFilterSeparator);
+                }
+            }
+            else
+            {
+                // Parse string range (for dates, text, etc.)
+                if (op[0] == '>')
+                {
+                    internalFilter.appendf("%u%c%s%c%s%c~~~~~~~~~~%c",
```
Comment (ghalliday) [inline plugins/fileservices/fileservices.cpp:2837]: Should there be a warning if (not?) hasEquals

File: plugins/fileservices/fileservices.cpp [hunk 7]
```diff
@@ -2385,6 +2391,652 @@ FILESERVICES_API void FILESERVICES_CALL fsLogicalFileList(ICodeContext *ctx, siz
...
+                        attrName.str(), DFUQFilterSeparator,
+                        DFUQFilterSeparator, value, DFUQFilterSeparator);
+                }
+            }
+            matched = true;
+        }
+        // Parse field=value as exact range match
+        else if (const char *eq = strchr(term, '='))
+        {
+            StringBuffer fieldName;
```
Comment (ghalliday) [inline plugins/fileservices/fileservices.cpp:2858]: Chunks of this are the same as the >< code above.  Potential to common up and simplify.  It may not be worth it...

File: plugins/fileservices/fileservices.cpp [hunk 8]
```diff
@@ -2385,6 +2391,652 @@ FILESERVICES_API void FILESERVICES_CALL fsLogicalFileList(ICodeContext *ctx, siz
...
+            internalFilter.appendf("%u%c%s%c%s%c",
+                DFUQFTwildcardMatch, DFUQFilterSeparator,
+                attrName.str(), DFUQFilterSeparator,
+                value, DFUQFilterSeparator);
+            matched = true;
+        }
+        // Parse field>value, field<value, field>=value, field<=value
+        else if (const char *op = strpbrk(term, "><"))
+        {
+            StringBuffer fieldName;
```
Comment (ghalliday) [inline plugins/fileservices/fileservices.cpp:2761]: Check for negate?

File: plugins/fileservices/fileservices.cpp [hunk 9]
```diff
@@ -2385,6 +2391,652 @@ FILESERVICES_API void FILESERVICES_CALL fsLogicalFileList(ICodeContext *ctx, siz
...
+        requestedFieldNames.append("cluster");
+    }
+    else
+    {
+        // Parse comma-separated field list
+        requestedFieldNames.appendList(requestedFields, ",");
+
+        // Ensure "name" is always included (required field)
+        bool hasName = false;
+        ForEachItemIn(idx, requestedFieldNames)
```
Comment (ghalliday) [inline plugins/fileservices/fileservices.cpp]: I think the following is equivalent:
```
hasName = requestedFieldNames.contains("name", true);
```

**Review by ghalliday — COMMENTED** [review]
@jakesmith looks like a very useful set of functionality.  A few comments.

**Review by ghalliday — CHANGES_REQUESTED** [review]
A few comments.

**Comment by jakesmith** [comment]
> Also add #ownwarning(2364, ignore);
> 
> to fix the regression suite mismatches.

I'll change the 2 tests over to new function (minor change).

File: testing/regress/ecl/logicalfilelist_filtered_test.ecl [hunk 10]
```diff
@@ -0,0 +1,285 @@
...
+END;
+
+// Test 7: Combined filters - record count with normal files (excludes super1 and large files)
+test7 := MODULE
+    EXPORT filters := 'rowcount>=200,rowcount<=500,is:normal';
+    EXPORT result := Std.File.LogicalFileListFiltered(prefix + '*', filters := filters);
+    EXPORT name := 'Test 7: Combined filters (200 <= rowcount <= 500 AND normal files)';
+    EXPORT verify := SEQUENTIAL(
+        OUTPUT(result.count, NAMED('Test7_Count')),
+        OUTPUT(result.count = 3, NAMED('Test7_ExpectedCount')), // Should match: file2(200), file3(500), testfile1(300), testfile2(400) = 4, but super1 excluded
```
[the remark this answers is not in the document]
Comment (jakesmith) [reply testing/regress/ecl/logicalfilelist_filtered_test.ecl]: agree. Will fix.

File: ecllibrary/std/File.ecl [hunk 11]
```diff
@@ -225,7 +225,92 @@ EXPORT dataset(FsFilenameRecord) RemoteDirectory(varstring machineIP, varstring
...
+    lib_fileservices.FileServices.LogicalFileList(namepattern, includenormal, includesuper, unknownszero, foreigndali)
+    : DEPRECATED('Use Std.File.LogicalFileListFiltered for enhanced filtering and count information');
+
+/**
+ * Returns information about logical files known to the system, including count and limit breach status.
+ * This function supports file limits, filtering, and field projection for efficient data retrieval.
+ *
+ * @param namepattern   The mask of the files to list. Defaults to '*' (all files).
+ * @param filters       Filter string using search-like syntax for complex queries. Defaults to blank.
+ *                      Multiple filters separated by spaces. Supports:
```
[the remark this answers is not in the document]
Comment (jakesmith) [reply ecllibrary/std/File.ecl]: will fix.

File: plugins/fileservices/fileservices.cpp [hunk 12]
```diff
@@ -2385,6 +2391,652 @@ FILESERVICES_API void FILESERVICES_CALL fsLogicalFileList(ICodeContext *ctx, siz
...
+                    if (!hasEquals)
+                    {
+                        if (maxVal == (-I64C(0x7FFFFFFFFFFFFFFF) - 1))
+                            throw makeStringExceptionV(-1, "Invalid filter syntax: '%s' - value too small for < comparison (would underflow)", originalTerm);
+                        maxVal--;
+                    }
+                    internalFilter.appendf("%u%c%s%c0%c%lld%c",
+                        DFUQFTinteger64Range, DFUQFilterSeparator,
+                        attrName.str(), DFUQFilterSeparator,
+                        DFUQFilterSeparator, maxVal, DFUQFilterSeparator);
```
[the remark this answers is not in the document]
Comment (jakesmith) [reply plugins/fileservices/fileservices.cpp:2825]: that's not true. It is currently sending an empty field value - which will be interpreted by the server as not having a lower bound to filter on.

File: plugins/fileservices/fileservices.cpp [hunk 13]
```diff
@@ -2385,6 +2391,652 @@ FILESERVICES_API void FILESERVICES_CALL fsLogicalFileList(ICodeContext *ctx, siz
...
+                        DFUQFTstringRange, DFUQFilterSeparator,
+                        attrName.str(), DFUQFilterSeparator,
+                        value, DFUQFilterSeparator, DFUQFilterSeparator);
+                }
+                else // op[0] == '<'
+                {
+                    internalFilter.appendf("%u%c%s%c%c%s%c",
+                        DFUQFTstringRange, DFUQFilterSeparator,
+                        attrName.str(), DFUQFilterSeparator,
+                        DFUQFilterSeparator, value, DFUQFilterSeparator);
```
[the remark this answers is not in the document]
Comment (jakesmith) [reply plugins/fileservices/fileservices.cpp:2849]: true, exclusive (> or <) isn't supported. Will clarify.

File: ecllibrary/std/File.ecl (hunk 1 above)
Comment (jakesmith) [reply ecllibrary/std/File.ecl]: it should be a 'remote' system, and not use a foreigndali. At the moment, it's the same as the LogicalFileList and only talks to Dali (there is no simple esp service wrapper so that a remote would work).

Foreign should be removed completely, forcing the use of remote if need be.
For now I'm going to change it to a remote storage name, but throw an unsupported exception, and cover in a future JIRA : https://hpccsystems.atlassian.net/browse/HPCC-35875

File: plugins/fileservices/fileservices.cpp (hunk 2 above)
Comment (jakesmith) [reply plugins/fileservices/fileservices.cpp:2665]: yeah, fixed.

File: plugins/fileservices/fileservices.cpp (hunk 3 above)
Comment (jakesmith) [reply plugins/fileservices/fileservices.cpp]: I think I'll leave this one as is.

Comment (jakesmith) [reply plugins/fileservices/fileservices.cpp]: actually, I'll change it.

File: plugins/fileservices/fileservices.cpp (hunk 4 above)
Comment (jakesmith) [reply plugins/fileservices/fileservices.cpp]: agree will change.

File: plugins/fileservices/fileservices.cpp (hunk 6 above)
Comment (jakesmith) [reply plugins/fileservices/fileservices.cpp:2837]: yeah, I'm changing it so only inclusive supported, will throw an exception if > or <

File: plugins/fileservices/fileservices.cpp (hunk 5 above)
Comment (jakesmith) [reply plugins/fileservices/fileservices.cpp:2802]: the server doesn't support floats.. I think I'll put a check in for the user filter being a float and reject it, but allow ints to be compared against fields that are float.

File: plugins/fileservices/fileservices.cpp (hunk 7 above)
Comment (jakesmith) [reply plugins/fileservices/fileservices.cpp:2858]: I think enough differences not to combine.

File: plugins/fileservices/fileservices.cpp (hunk 8 above)
Comment (jakesmith) [reply plugins/fileservices/fileservices.cpp:2761]: I'll add a check, they're not supported on most things at the moment.

File: plugins/fileservices/fileservices.cpp (hunk 9 above)
Comment (jakesmith) [reply plugins/fileservices/fileservices.cpp]: agreed.

**Comment by jakesmith** [comment]
@ghalliday - please see changes in 2nd commit.

workflow_9c.ecl failed [unrelated]. Should be disable it (at least for now), rather than live with intermittent unrelated failures like this?
```
[Failure]  59. Fail workflow_9c.ecl ( version: parallel=true ) - W20260219-114259 (1603 sec)
[Failure]  59. URL http://127.0.0.1:8010/?Widget=WUDetailsWidget&Wuid=W20260219-114259
[Failure]  59. Create Stack Trace for roxie (pid:7917) into '/home/runner/HPCCSystems-regression/log/W20260219-114259-roxie.trace'
[Failure]  59. Zipped Analysis Package: ZAP file written to /home/runner/HPCCSystems-regression/zap/ZAPReport_W20260219-114259_regress.zip.
    
[Failure]  59. Execute pre abort script '/opt/HPCCSystems/bin/smoketest-preabort.sh'
[Failure]  59. Pre abort script result stored into '/home/runner/HPCCSystems-regression/log/W20260219-114259-preAbort.log'
[Failure]  59. Timeout occured for workflow_9c.ecl and no more attempt left. Force to abort... 
Suite destructor.

**Review by ghalliday — APPROVED** [review]
Looks good, please squash

**Comment by jakesmith** [comment]
@ghalliday - squashed.
